src/js/heuristicblocking.js

```diff
@@ -499,38 +445,96 @@ function hasCookieTracking(details, origin) {
   return false;
 }
 
-function startListeners() {
-  /**
-   * Adds heuristicBlockingAccounting as listened to onBeforeSendHeaders request
-   */
-  chrome.webRequest.onBeforeSendHeaders.addListener(function(details) {
-    if (badger) {
-      return badger.heuristicBlocking.heuristicBlockingAccounting(details);
-    } else {
-      return {};
-    }
-  }, {urls: ["<all_urls>"]}, ["requestHeaders"]);
+/**
+ * Process a webrequest for use by heuristicblocking accounting. If the request
+ * isn't relevant then we return early. Optionally check the origin for DNT.
+ *
+ * @param details from onBeforeSendHeaders or onBeforeResponseStarted
+ * @param callback heuristicBlockingAccounting
+ * @param checkDNT bool whether to check the origin for DNT
+ * @ returns {*}
+ */
+function _processWebRequest(details, callback, checkDNT) {
+  // ignore requests that not from tabs, or are from incognito tabs
+  if (details.tabId < 0 || incognito.tabIsIncognito(details.tabId)) {
+    return {};
+  }
 
-  /**
-   * Adds onResponseStarted listener. Monitor for cookies
-   */
-  chrome.webRequest.onResponseStarted.addListener(function(details) {
-    var hasSetCookie = false;
-    for(var i = 0; i < details.responseHeaders.length; i++) {
+  let fqdn = utils.makeURI(details.url).host,
+    baseDomain = window.getBaseDomain(fqdn);
+
+  // if this is a main window request aka first party origin
+  if (details.type == "main_frame") {
+    // save the origin associated with the tab
+    log("Origin: " + baseDomain + "\tURL: " + details.url);
+    tabOrigins[details.tabId] = baseDomain;
+    return {};
+  }
+
+  let tabOrigin = tabOrigins[details.tabId];
+  // ignore first-party requests
+  if (!tabOrigin || baseDomain == tabOrigin) {
+    return {};
+  }
+
+  if (checkDNT) {
+    badger.checkForDNTPolicy(fqdn);
+  }
+
+  return callback(details, fqdn, baseDomain, tabOrigin);
+}
+
+/**
+ * Asynchronous onBeforeSendHeaders listener.
+ */
+function onBeforeSendHeaders(details) {
+  return _processWebRequest(details, heuristicBlockingAccounting, true);
+}
+
+function onBeforeResponseStarted(details) {
```

Comment:
`onBeforeResponseStarted` -> `onResponseStarted` so as to match the webRequest API. Might be removed in a future PR, though in case it doesn't we should fix it here.